packages/react/src/domain/handles.ts

```diff
@@ -26,15 +28,28 @@ const FACTORS_WITH_EMAIL = [
 const FACTORS_WITH_PHONE = ["otp_via_sms", "sms_link", "password"];
 const SSO_FACTORS = ["oidc", "saml"];
 
-function getPossibleHandleTypes(factor: Factor): Set<HandleType> {
+function getPossibleHandleTypes(
+  factor: Factor | FactorWithAllowedHandleTypes
+): Set<HandleType> {
   const handleTypes = new Set<HandleType>();
 
   if (FACTORS_WITH_EMAIL.includes(factor.method)) {
-    handleTypes.add("email_address");
+    if (
+      !isFactorWithAllowedHandleTypes(factor) ||
+      (isFactorWithAllowedHandleTypes(factor) &&
+        factor.allowedHandleTypes?.includes("email_address"))
+    )
+      handleTypes.add("email_address");
   }
 
   if (FACTORS_WITH_PHONE.includes(factor.method)) {
-    handleTypes.add("phone_number");
+    if (
```

Comment:
There's a bit of duplication now, how do you feel about extracting the new check in a separate function?

`function isHandleTypeAllowed(factor: Factor | FactorWithAllowedHandleTypes, handleType: HandleType): boolean` could help make this more readable